agents/magpie/magpie_agent.py

```diff
@@ -0,0 +1,743 @@
+        nsamps = len(times_in)
+        nchans = len(data_in)
+
+        if self.avg1 is None:
+            self.avg1 = RollingAvg(200, nchans)
+            self.avg2 = RollingAvg(200, nchans)
+        elif self.avg1.nchans != nchans:
+            self.log.warn(f"Channel count has changed! {self.avg1.nchans}->{nchans}")
+            self.avg1 = RollingAvg(200, nchans)
+            self.avg2 = RollingAvg(200, nchans)
+
+        # Calc RMS
+        hpf_data = data_in - self.avg1.apply(data_in)  # To high-pass filter
+        rms = np.sqrt(self.avg2.apply(hpf_data**2))  # To calc rolling rms
+
+        ds_factor = sample_rate // self.target_rate
+        if np.isnan(ds_factor):  # There is only one element in the timestream
+            ds_factor = 1
+        ds_factor = max(int(ds_factor), 1)  # Prevents downsample factors < 1
+
+        # Arrange output data structure
+        sample_idxs = np.arange(0, nsamps, ds_factor, dtype=np.int32)
+        num_frames = len(sample_idxs)
+
+        times_out = times_in[sample_idxs]
+
+        abs_chans = self.mask[np.arange(nchans)]
+
+        raw_out = np.zeros((num_frames, self.fp.num_dets))
+        rms_out = np.zeros((num_frames, self.fp.num_dets))
+
+        for i, c in enumerate(abs_chans):
+            if c >= len(self.fp.chan_mask):
+                continue
+            idx = self.fp.chan_mask[c]
+            if idx >= 0:
+                raw_out[:, idx] = data_in[i, sample_idxs]
+                rms_out[:, idx] = rms[i, sample_idxs]
+
+        out = []
+        for i in range(num_frames):
+            fr = core.G3Frame(core.G3FrameType.Scan)
+            fr['idx'] = 0
+            fr['data'] = core.G3VectorDouble(raw_out[i])
+            fr['timestamp'] = core.G3Time(times_out[i] * core.G3Units.s)
+            out.append(fr)
+
+            fr = core.G3Frame(core.G3FrameType.Scan)
+            fr['idx'] = 1
+            fr['data'] = core.G3VectorDouble(rms_out[i])
+            fr['timestamp'] = core.G3Time(times_out[i] * core.G3Units.s)
+            out.append(fr)
+        return out
+
+
+    def read(self, session, params=None):
+        """read(src='tcp://localhost:4532')
+
+        **Process** - Process for reading in G3Frames from a source or list of
+        sources. If this source is an address that begins with ``tcp://``, the
+        agent will attempt to connect to a G3NetworkSender at the specified
+        location. The ``src`` param can also be a filepath or list of filepaths
+        pointing to G3Files to be streamed. If a list of filenames is passed,
+        once the first file is finished streaming, subsequent files will be
+        streamed.
+        """
+
+        self._running = True
+        session.set_status('running')
+
+        src_idx = 0 
+        if isinstance(params['src'], str):
+            sources = [params['src']]
+        else:
+            sources = params['src']
+
+        reader = core.G3Reader(sources[src_idx])
```

Comment:
What happens if this can't connect to the `tcp://` address?

Reply:
Currently the agent will just crash. I was planning on eventually making this more robust by adding the logic from the smurf-recorder agent, but wasn't originally planning on doing it in this PR.... I think it does make sense to add it for robustness so I'll try it out.